src/main/java/net/fabricmc/loom/configuration/providers/forge/MinecraftPatchedProvider.java

```diff
@@ -412,20 +466,42 @@ private void remapPatchedJar() throws Exception {
 			remapper.finish();
 		}
 
-		copyUserdevFiles(forgeUserdevJar, minecraftPatchedSrgJar);
+		copyUserdevFiles(forgeUserdevJar, output);
+
+		Files.copy(output, input, StandardCopyOption.REPLACE_EXISTING);
+	}
+
+	private void remapPatchedJar() throws Exception {
+		logger.lifecycle(":remapping minecraft (TinyRemapper, srg -> official)");
+		Path mcInput = minecraftPatchedSrgAtJar;
+		Path mcOutput = minecraftPatchedJar;
+		Files.deleteIfExists(mcOutput);
+
+		TinyRemapper remapper = buildRemapper(mcInput);
+
+		try (OutputConsumerPath outputConsumer = new OutputConsumerPath.Builder(mcOutput).build()) {
+			outputConsumer.addNonClassFiles(mcInput);
+			remapper.readInputs(mcInput);
+			remapper.apply(outputConsumer);
+		} finally {
+			remapper.finish();
+		}
+
 		applyLoomPatchVersion(mcOutput);
 	}
 
-	private void patchJars() throws Exception {
+	private void patchJars(Path input, Path output) throws Exception {
 		Stopwatch stopwatch = Stopwatch.createStarted();
 		logger.lifecycle(":patching jars");
-		patchJars(minecraftSrgJar, minecraftPatchedSrgJar, type.patches.apply(getExtension().getPatchProvider(), getExtension().getForgeUserdevProvider()));
+		patchJars(input, output, type.patches.apply(getExtension().getPatchProvider(), getExtension().getForgeUserdevProvider()));
+
+		mergeForge(output);
```

Comment:
It breaks on older versions as the hierarchy for `ShapedRecipe` relies on the forge class `IShapedRecipe`, so remapping doesn't produce correct results since there's a missing class in between the base recipe and shaped recipe classes.